src/components/Table/utils/TableState.ts

```diff
@@ -218,25 +218,61 @@ export class TableState {
-      this.columns = columns;
       this.visibleColumnsStorageKey = visibleColumnsStorageKey ?? camelCase(columns.map((c) => c.id).join());
       this.visibleColumns.replace(readOrSetLocalVisibleColumnState(columns, this.visibleColumnsStorageKey));
-      const expandedColumnIds = columns.filter((c) => c.initExpanded).map((c) => c.id);
-      if (isInitial) this.expandedColumns.replace(expandedColumnIds);
+      // list of local storage columns
+      const localStorageColumns = this.persistCollapse ? readExpandedColumnsStorage(this.persistCollapse) : [];
+      // expandedColumnIds initialized as localStorageColumns
+      const existingExpandedColumnIds = isInitial ? localStorageColumns : this.expandedColumnIds;
+
+      // list of all existing columns
+      const existingColumnIds = this.columns.map((c) => c.id);
+
+      // filter existing columns that are initExpanded to see which are new
+      const newExpandedColumnsIds = columns
+        .filter((c) => !existingColumnIds.includes(c.id) && c.initExpanded)
+        .map((c) => c.id);
+
+      const isDifferent = !newExpandedColumnsIds.every((c) => existingExpandedColumnIds.includes(c));
+      if (isDifferent) {
+        existingExpandedColumnIds.push(...newExpandedColumnsIds);
+      }
+
+      const newExpanded = columns.filter((c) => existingExpandedColumnIds.includes(c.id));
+      // if we have existing expanded columns ids, then do a Promise.all() and execute column.expandColumns
+      Promise.all(
+        newExpanded.map(async (c) => {
+          await this.loadExpandedColumns(c);
+        }),
+      ).then(() => {
+        const newIds = newExpanded.map((c) => c.id);
+        // if there is a difference between list of current expanded columns vs list we just created, then replace
+
+        if (isInitial || isDifferent) {
+          this.expandedColumns.replace(newIds);
+        }
+        //  update our persistCollapse if set
+        if (this.persistCollapse) {
+          sessionStorage.setItem(`expandedColumn_${this.persistCollapse}`, JSON.stringify(newIds));
+        }
+      });
+      this.columns = columns;
     }
   }
 
   // load and trigger column to be expanded
   async loadExpandedColumns(column: GridColumnWithId<any>): Promise<void> {
-    // if we dont have anything in our cache and our expanded columns are a function
+    // if we don't have anything in our cache and our expanded columns are a function
     if (!this.loadedColumns.has(column.id) && isFunction(column.expandColumns)) {
       // set our result to the function call of expandColumns
       const result = await column.expandColumns();
       // once we have the loaded columns, add result to local cache
       this.loadedColumns.set(column.id, assignDefaultColumnIds(result));
     }
-    // once column is in local cache, then toggle
-    this.toggleExpandedColumn(column.id);
+    if (column.initExpanded) {
+      // manually toggle expanded column
+      this.toggleExpandedColumn(column.id);
+    }
```

Comment:
**thought/question:** Above there is a code change & comment that says: 
```
await tableState.loadExpandedColumns(column);
// manually calling this as loadExpandedColumns does not toggle
tableState.toggleExpandedColumn(column.id);
```
But it looks like `loadExpandedColumns` does call `toggle` for a column that is `initExpanded`. 
I'm thinking we probably do not want this `toggleExpandedColumn` within `loadExpandedColumns` like your comment suggests. Otherwise for columns that are `initExpanded`, when a user clicks the ExpandableHeader component, then I'm pretty sure we'd call `toggleExpandedColumn` twice. Once in `ExpandableHeader` and another time here. I'd think that'd result not being able to change the collapsed state of the column.

Which code path was this logic created for?

Reply:
Gotcha. I feel like there is a way to simplify the logic here. Will think about it/play around with it.